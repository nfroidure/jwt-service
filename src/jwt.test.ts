// @ts-ignore: no type atm  ¯\_(ツ)_/¯
import YError from 'yerror';
import initJWTService from './jwt';

describe('jwt service', () => {
  const log = jest.fn();
  const time = jest.fn();

  afterEach(() => {
    log.mockReset();
    time.mockReset();
  });

  describe('initializer', () => {
    test('should fail without secret', async () => {
      try {
        await initJWTService({
          JWT: {
            duration: '2d',
            tolerance: '2h',
            algorithms: ['HS256'],
          },
          log,
          time,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail without algorithms', async () => {
      try {
        await initJWTService({
          ENV: {
            JWT_SECRET: 'test',
          },
          JWT: {
            duration: '2d',
            tolerance: '2h',
            algorithms: [],
          },
          log,
          time,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail without duration', async () => {
      try {
        await initJWTService({
          ENV: {
            JWT_SECRET: 'test',
          },
          // @ts-ignore: no type atm ¯\_(ツ)_/¯
          JWT: {
            tolerance: '2h',
            algorithms: ['HS256'],
          },
          log,
          time,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail with a bad tolerance', async () => {
      try {
        await initJWTService({
          ENV: {
            JWT_SECRET: 'test',
          },
          JWT: {
            duration: '2h',
            tolerance: '',
            algorithms: ['HS256'],
          },
          log,
          time,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail with a uninterpreted duration', async () => {
      try {
        await initJWTService({
          ENV: {
            JWT_SECRET: 'test',
          },
          JWT: {
            duration: 'q',
            algorithms: ['HS256'],
          },
          log,
          time,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fallback to default tolerance', async () => {
      await initJWTService({
        ENV: {
          JWT_SECRET: 'test',
        },
        JWT: {
          duration: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });
    });
  });

  describe('sign', () => {
    test('should work', async () => {
      time.mockReturnValueOnce(new Date('2014-01-26T00:00:00Z').getTime());

      const jwt = await initJWTService({
        ENV: {
          JWT_SECRET: 'secret',
        },
        JWT: {
          duration: '2d',
          tolerance: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });
      const token = await jwt.sign({
        userId: 2,
        organisationId: 3,
      });

      expect({
        token,
        logs: log.mock.calls,
        times: time.mock.calls,
      }).toMatchSnapshot();
    });

    test('should fail with a bad algorithm', async () => {
      time.mockReturnValueOnce(new Date('2014-01-26T00:00:00Z').getTime());

      const jwt = await initJWTService({
        JWT: {
          secret: 'secret',
          duration: '2d',
          tolerance: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });

      try {
        await jwt.sign(
          {
            userId: 2,
            organisationId: 3,
          },
          'LOLALG',
        );
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });

  describe('verify', () => {
    test('should work', async () => {
      time.mockReturnValueOnce(new Date('2014-01-26T00:00:00Z').getTime());

      const jwt = await initJWTService({
        JWT: {
          secret: 'secret',
          duration: '2d',
          tolerance: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });
      const decoded = await jwt.verify(
        'eyJhbGciOiJIUzI1NiJ9.' +
          'eyJ1c2VySWQiOjIsIm9yZ2FuaXNhdGlvbklkIjozLCJpY' +
          'XQiOjEzOTA2OTQ0MDAsImV4cCI6MTM5MDg2NzIwMCwibmJmIjoxMzkwNjk0NDAwfQ.' +
          'DdWhIErffR-N-bTSsjr2tDOyinbMtYkL24IZxOVaB_0',
      );

      expect({
        decoded,
        logs: log.mock.calls,
        times: time.mock.calls,
      }).toMatchSnapshot();
    });

    test('should fail after the validity duration', async () => {
      time.mockReturnValueOnce(new Date('2014-03-30T00:00:00Z').getTime());

      const jwt = await initJWTService({
        JWT: {
          secret: 'secret',
          duration: '2d',
          tolerance: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });

      try {
        await jwt.verify(
          'eyJhbGciOiJIUzI1NiJ9.' +
            'eyJ1c2VySWQiOjIsIm9yZ2FuaXNhdGlvbklkIjozLCJpY' +
            'XQiOjEzOTA2OTQ0MDAsImV4cCI6MTM5MDg2NzIwMCwibmJmIjoxMzkwNjk0NDAwfQ.' +
            'DdWhIErffR-N-bTSsjr2tDOyinbMtYkL24IZxOVaB_0',
        );
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail with a malformed token', async () => {
      time.mockReturnValueOnce(new Date('2014-03-30T00:00:00Z').getTime());

      const jwt = await initJWTService({
        JWT: {
          secret: 'secret',
          duration: '2d',
          tolerance: '2h',
          algorithms: ['HS256'],
        },
        log,
        time,
      });

      try {
        await jwt.verify('kikooolol');
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logs: log.mock.calls,
          times: time.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
