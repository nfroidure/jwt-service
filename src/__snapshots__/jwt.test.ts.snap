// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jwt service initializer should fail with a bad tolerance 1`] = `
Object {
  "errorCode": "E_BAD_JWT_TOLERANCE",
  "errorParams": Array [
    "",
    "val is not a non-empty string or a valid number. val=\\"\\"",
  ],
  "logs": Array [],
  "times": Array [],
}
`;

exports[`jwt service initializer should fail with a uninterpreted duration 1`] = `
Object {
  "errorCode": "E_BAD_JWT_DURATION",
  "errorParams": Array [
    "q",
  ],
  "logs": Array [],
  "times": Array [],
}
`;

exports[`jwt service initializer should fail without algorithms 1`] = `
Object {
  "errorCode": "E_NO_JWT_ALGORITHMS",
  "errorParams": Array [],
  "logs": Array [],
  "times": Array [],
}
`;

exports[`jwt service initializer should fail without duration 1`] = `
Object {
  "errorCode": "E_BAD_JWT_DURATION",
  "errorParams": Array [],
  "logs": Array [],
  "times": Array [],
}
`;

exports[`jwt service initializer should fail without secret 1`] = `
Object {
  "errorCode": "E_NO_JWT_SECRET",
  "errorParams": Array [],
  "logs": Array [],
  "times": Array [],
}
`;

exports[`jwt service sign should fail with a bad algorithm 1`] = `
Object {
  "errorCode": "E_UNKNOWN_ALGORYTHM",
  "errorParams": Array [
    "LOLALG",
    Array [
      "HS256",
    ],
  ],
  "logs": Array [
    Array [
      "info",
      "JWT service initialized!",
    ],
  ],
  "times": Array [
    Array [],
  ],
}
`;

exports[`jwt service sign should work 1`] = `
Object {
  "logs": Array [
    Array [
      "info",
      "JWT service initialized!",
    ],
  ],
  "times": Array [
    Array [],
  ],
  "token": Object {
    "expiresAt": 1390867200000,
    "issuedAt": 1390694400000,
    "token": "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOjIsIm9yZ2FuaXNhdGlvbklkIjozLCJpYXQiOjEzOTA2OTQ0MDAsImV4cCI6MTM5MDg2NzIwMCwibmJmIjoxMzkwNjk0NDAwfQ.DdWhIErffR-N-bTSsjr2tDOyinbMtYkL24IZxOVaB_0",
    "validAt": 1390694400000,
  },
}
`;

exports[`jwt service verify should fail after the validity duration 1`] = `
Object {
  "errorCode": "E_JWT_EXPIRED",
  "errorParams": Array [
    "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOjIsIm9yZ2FuaXNhdGlvbklkIjozLCJpYXQiOjEzOTA2OTQ0MDAsImV4cCI6MTM5MDg2NzIwMCwibmJmIjoxMzkwNjk0NDAwfQ.DdWhIErffR-N-bTSsjr2tDOyinbMtYkL24IZxOVaB_0",
    "jwt expired",
  ],
  "logs": Array [
    Array [
      "info",
      "JWT service initialized!",
    ],
  ],
  "times": Array [
    Array [],
  ],
}
`;

exports[`jwt service verify should fail with a malformed token 1`] = `
Object {
  "errorCode": "E_JWT_MALFORMED",
  "errorParams": Array [
    "kikooolol",
    "jwt malformed",
  ],
  "logs": Array [
    Array [
      "info",
      "JWT service initialized!",
    ],
  ],
  "times": Array [
    Array [],
  ],
}
`;

exports[`jwt service verify should work 1`] = `
Object {
  "decoded": Object {
    "exp": 1390867200,
    "iat": 1390694400,
    "nbf": 1390694400,
    "organisationId": 3,
    "userId": 2,
  },
  "logs": Array [
    Array [
      "info",
      "JWT service initialized!",
    ],
  ],
  "times": Array [
    Array [],
  ],
}
`;
